File: linked_list/cpp/double_linkedlist.cpp
#include <iostream>
#include <cstdlib>

typedef int element;

struct Node {
    element data;
    Node* next;
    Node* prev;
};

struct LList {
    int size;
    Node *head;
    Node *tail;
};

LList *create_list() {
    LList *list = new LList;
    list->size = 0;
    list->head = nullptr;
    list->tail = nullptr;
    return list;
}

int get_size(LList* list) {
    return list->size;
}

bool is_empty(LList* list) {
    return list->size == 0;
}

void clear(LList* list) {
    Node* current = list->head;
    Node* next_node;

    while (current != nullptr) {
        next_node = current->next;
        delete current;
        current = next_node;
    }

    list->head = list->tail = nullptr;
    list->size = 0;
}

int get_first(LList *list) {
    if (list->head != nullptr) {
        return list->head->data;
    }
    return -1;
}

int get_last(LList * list) {
    if (list->tail != nullptr) {
        return list->tail->data;
    }
    return -1;
}

void print_list(LList *list) {
    Node *cur = list->head;
    while (cur != nullptr) {
        std::cout << cur->data << " ";
        cur = cur->next;
    }
    std::cout << std::endl;
}

void add_first(LList * list, element item) {
    Node *temp = new Node;
    temp->data = item;
    temp->next = list->head;
    temp->prev = nullptr;
    if (list->head != nullptr) {
        list->head->prev = temp;
    }
    else {
        list->tail = temp;
    }
    list->head = temp;
    list->size++;
}

void add_before(LList *list, Node* next_node, element item) {
    if (next_node == nullptr) return;
    Node *temp = new Node;
    temp->data = item;
    temp->prev = next_node->prev;
    temp->next = next_node;
    if (next_node->prev != nullptr) {
        next_node->prev->next = temp;
    }
    else {
        list->head = temp;
    }
    next_node->prev = temp;
    list->size++;
}

void add_after(LList* list, Node* prev_node, int item) {
    if (prev_node == nullptr) return;

    Node* temp = new Node;
    temp->data = item;
    temp->next = prev_node->next;
    temp->prev = prev_node;

    if (prev_node->next != nullptr) {
        prev_node->next->prev = temp;
    } else {
        list->tail = temp;
    }

    prev_node->next = temp;
    list->size++;
}

Node* get_node_by_index(LList *list, int index) {
    if (index < 0) return nullptr;
    Node *p = list->head;
    for (int i = 0; i < index && p != nullptr; i++) {
        p = p->next;
    }
    return p;
}

void add_by_index(LList* list, int index, element item) {
    if (index < 0 || index > list->size) {
        return; // Handle out-of-bounds index
    }

    if (index == 0) {
        add_first(list, item);
        return;
    }

    Node *prev = get_node_by_index(list, index - 1);
    if (prev != nullptr) {
        add_after(list, prev, item);
    }
}

void add_to_ordered_list(LList *list, element item) {
    Node *p = list->head;
    Node *q = nullptr;

    while (p != nullptr && p->data < item) {
        q = p;
        p = p->next;
    }

    if (q == nullptr) {
        add_first(list, item);
    } else {
        add_after(list, q, item);
    }
}

Node* find_node(LList *list, element item) {
    Node *cur = list->head;
    while (cur != nullptr) {
        if (cur->data == item){
            return cur;
        }
        cur = cur -> next;
    }
    return nullptr;
}

void remove_first(LList *list) {
    if (list->head == nullptr) return;
    Node *temp = list->head;
    list->head = list->head->next;
    if (list->head != nullptr) {
        list->head->prev = nullptr;
    }
    delete temp;
    list->size--;
}

void remove_last(LList *list) {
    if (list->tail == nullptr) return;
    Node *temp = list->tail;
    list->tail = list->tail->prev;
    if (list->tail != nullptr) {
        list->tail->next = nullptr;
    }
    else{
        list->head = nullptr;
    }
    delete temp;
    list->size--;
}

void remove_after(LList* list, Node* prev_node) {
    if (prev_node == nullptr || prev_node->next == nullptr) return;

    Node* temp = prev_node->next;
    prev_node->next = temp->next;

    if (temp->next != nullptr) {
        temp->next->prev = prev_node;
    } else {
        list->tail = prev_node;
    }

    delete temp;
    list->size--;
}

void remove_by_index(LList* list, int index) {
    if (index < 0) {
        return;
    }

    if (index == 0) {
        remove_first(list);
    }

    Node *prev = get_node_by_index(list, index-1);
    if (prev != nullptr && prev->next != nullptr) {
        remove_after(list, prev);
    }
}

void remove_by_value(LList* list, element item) {
    Node* node_to_remove = find_node(list, item);

    if (node_to_remove == nullptr) return;

    if (node_to_remove == list->head) {
        remove_first(list);
    } else if (node_to_remove == list->tail) {
        remove_last(list);
    } else {
        Node* prev = node_to_remove->prev;
        remove_after(list, prev);
    }
}

Node* merge(Node *list1, Node *list2) {
    Node dummy;
    Node *tail = &dummy;
    dummy.next = nullptr;

    while (list1 != nullptr && list2 != nullptr) {
        if (list1->data <= list2->data) {
            tail->next = list1;
            list1->prev = tail;
            list1 = list1->next;
        } else {
            tail->next = list2;
            list2->prev = tail;
            list2 = list2->next;
        }
        tail = tail->next;
    }
    tail->next = (list1 != nullptr) ? list1 : list2;
    if (tail->next != nullptr) {
        tail->next->prev = tail;
    }
    return dummy.next;
}

void concat(LList* list1, LList* list2) {
    if (list1->head == nullptr) {
        list1->head = list2->head;
        list1->tail = list2->tail;
    } else if (list2->head != nullptr) {
        list1->tail->next = list2->head;
        list2->head->prev = list1->tail;
        list1->tail = list2->tail;
    }
    list1->size += list2->size;
    list2->head = list2->tail = nullptr;
    list2->size = 0;
}

Node* get_mid(LList* list) {
    if (list == nullptr || list->head == nullptr) return nullptr;
    Node* slow = list->head;
    Node* fast = list->head;

    while (fast->next != nullptr && fast->next->next != nullptr) {
        slow = slow->next;
        fast = fast->next->next;
    }
    return slow;
}

void sort(LList *list) {
    if (list == nullptr || list->head == nullptr || list->head->next == nullptr) {
        return; // Base case for recursion
    }

    Node *mid = get_mid(list);
    if (mid == nullptr || mid->next == nullptr) return; // Safety check

    Node *half = mid->next;
    mid->next = nullptr; // Split the list into two halves
    if (half) {
        half->prev = nullptr; // Ensure the start of the second list has no previous node
    }

    LList* left_sorted = create_list();
    LList* right_sorted = create_list();

    left_sorted->head = list->head;
    left_sorted->tail = mid;
    left_sorted->size = list->size / 2;

    right_sorted->head = half;
    right_sorted->tail = list->tail;
    right_sorted->size = list->size - left_sorted->size;

    sort(left_sorted);
    sort(right_sorted);

    list->head = merge(left_sorted->head, right_sorted->head);

    Node* cur = list->head;
    while (cur && cur->next) {
        cur = cur->next;
    }
    list->tail = cur;
    list->size = left_sorted->size + right_sorted->size;

    delete left_sorted;
    delete right_sorted;
}

// Insertion sort for the linked list
void insertion_sort(LList* list) {
    if (!list->head || !list->head->next) return;

    Node* sorted_head = list->head;
    Node* cur = list->head->next;
    sorted_head->next = nullptr;

    while (cur) {
        Node* next_cur = cur->next;
        if (cur->data < sorted_head->data) {
            cur->next = sorted_head;
            sorted_head->prev = cur;
            sorted_head = cur;
        } else {
            Node* search = sorted_head;
            while (search->next && search->next->data < cur->data) {
                search = search->next;
            }
            cur->next = search->next;
            if (search->next) {
                search->next->prev = cur;
            }
            search->next = cur;
            cur->prev = search;
        }
        cur = next_cur;
    }

    list->head = sorted_head;
    list->tail = sorted_head;
    while (list->tail && list->tail->next) {
        list->tail = list->tail->next;
    }
}

// Selection sort for the linked list
void selection_sort(LList* list) {
    if (!list->head) return;

    Node* cur = list->head;
    while (cur) {
        Node* min_node = cur;
        Node* search = cur->next;
        while (search) {
            if (search->data < min_node->data) {
                min_node = search;
            }
            search = search->next;
        }
        if (min_node != cur) {
            std::swap(cur->data, min_node->data);
        }
        cur = cur->next;
    }
}


void reverse(LList* list) {
    Node* prev = nullptr;
    Node* cur = list->head;
    Node* next = nullptr;

    while (cur) {
        next = cur->next;
        cur->next = prev;
        cur->prev = next;
        prev = cur;
        cur = next;
    }

    list->tail = list->head;
    list->head = prev;
}

void rotate(LList* list, int d) {
    if (list->head == nullptr || d == 0 || list->size == 0) return;

    d = d % list->size;
    if (d == 0) return;

    if (d < 0) d = list->size + d;

    Node* new_tail = list->head;
    for (int i = 0; i < list->size - d - 1; i++) {
        new_tail = new_tail->next;
    }

    Node* new_head = new_tail->next;
    new_tail->next = nullptr;
    new_head->prev = nullptr;

    list->tail->next = list->head;
    list->head->prev = list->tail;

    list->head = new_head;
    list->tail = new_tail;
}


///// array_to_linkedlist에 사용예정
void add_last(LList* list, element item) {
    Node* temp = new Node;
    temp->data = item;
    temp->next = nullptr;
    temp->prev = list->tail;

    if (list->tail != nullptr) {
        list->tail->next = temp;
    } else {
        list->head = temp;  // 리스트가 비어있을 경우 head도 갱신합니다.
    }

    list->tail = temp;
    list->size++;
}

void array_to_linkedlist(LList* list, element arr[], int size) {
    clear(list);  // 기존 리스트를 초기화합니다.

    for (int i = 0; i < size; ++i) {
        add_last(list, arr[i]);  // 배열의 각 요소를 리스트의 끝에 추가합니다.
    }
}

int* linkedlist_to_array(LList* list) {
    if (list->size == 0) return nullptr;  // 빈 리스트는 nullptr을 반환합니다.

    int* arr = new int[list->size];  // 리스트 크기만큼 배열을 동적 할당합니다.
    Node* current = list->head;
    int index = 0;

    while (current != nullptr) {
        arr[index++] = current->data;  // 현재 노드의 데이터를 배열에 복사합니다.
        current = current->next;
    }

    return arr;
}

int main() {
    LList* llist = create_list();

    // Initial list setup
    add_first(llist, 3); // Thursday
    add_first(llist, 2); // Tuesday
    add_first(llist, 1); // Monday

    std::cout << "Initial List:" << std::endl;
    print_list(llist);

    // Reverse the list
    reverse(llist);
    std::cout << "Reversed List:" << std::endl;
    print_list(llist);
    
    // Insertion sort the list
    insertion_sort(llist);
    std::cout << "Insertion sort List:" << std::endl;
    print_list(llist);

    // Add to ordered list
    add_to_ordered_list(llist, 4); // Wednesday
    std::cout << "Ordered Insertion:" << std::endl;
    print_list(llist);

    rotate(llist, 2);
    std::cout << "List after rotating by 2 positions:" << std::endl;
    print_list(llist);

    // Merge two lists
    LList* list1 = create_list();
    LList* list2 = create_list();

    add_first(list1, 4); // D
    add_first(list1, 2); // B
    add_first(list1, 1); // A

    add_first(list2, 5); // E
    add_first(list2, 3); // C

    // Node *merged_list = merge(sort(list1), sort(list2));
    // printf("Merged and Sorted List:\n");
    // print_list(merged_list);

    // Concatenate two lists
    concat(list1, list2);
    std::cout << "Concatenated List:" << std::endl;
    print_list(list1);
   
    // Reverse the list
    reverse(list1);
    std::cout << "Reversed List:" << std::endl;
    print_list(list1);

    // Sort the list
    selection_sort(list1);
    std::cout << "Selection sort List:" << std::endl;
    print_list(list1);

     // LinkedList to Array
    int* array_from_list = linkedlist_to_array(list1);
    std::cout << "Array after converting from linked list:" << std::endl;
    for (int i = 0; i < get_size(list1); ++i) {
        std::cout << array_from_list[i] << " ";
    }
    std::cout << std::endl;

    // Array to Linkedlist
    LList* new_list = create_list();
    array_to_linkedlist(new_list, array_from_list, list1->size);
    std::cout << "New Linked List converted from array:" << std::endl;
    print_list(new_list);

    return 0;
}

File: double_linkedlist.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        self.prev = None


class LList:
    def __init__(self):
        self.size = 0
        self.head = None
        self.tail = None

    #Print the size
    def get_size(self):
        return self.size
    #만약 위에 def __init__(self) 함수 없으면 def get_size(head) 해야함

    def is_empty(self):
        return self.size == 0
    
    def clear(self):
        self.head = self.tail = None
        self.size = 0

    def get_first(self):
        return self.head.data if self.head else None

    def get_last(self):
        return self.tail.data if self.tail else None
    

    # Print the linked list
    def print_list(self):
        cur = self.head
        while cur:
            print(cur.data, end=" ")
            cur = cur.next
        print()

    # Insert node at the beginning of the list
    # def add_first(head, item):
    #     temp = Node(item)
    #     temp.next = head
    #     if head:
    #         head.prev = temp
    #     return temp  # New head node's address
    
    # 혹은, self가 필요하다면
    def add_first(self, item):
        temp = Node(item)
        temp.next = self.head
        if self.head:
            self.head.prev = temp
        else:
            self.tail = temp
        self.head = temp
        self.size += 1
    
    #Insert node before a specific node
    # def add_before(next, item):
    #     if next is None:
    #         return
    #     temp = Node(item)
    #     temp.prev = next.prev
    #     temp.next = next
    #     if next.prev:
    #         next.prev.next = temp
    #     next.prev = temp

    #혹은 self를 쓰고싶다면
    def add_before(self, next_node, item):
        if next_node is None:
            return
        temp = Node(item)
        temp.prev = next_node.prev
        temp.next = next_node
        if next_node.prev:
            next_node.prev.next = temp
        else:  # next_node was the head
            self.head = temp
        next_node.prev = temp
        self.size += 1

    # Insert node after a specific node
    def add_after(self, prev, item):
        if prev is None:
            return
        temp = Node(item)
        temp.next = prev.next
        temp.prev = prev
        if prev.next:
            prev.next.prev = temp
        else:
            self.tail = temp
        prev.next = temp
        self.size += 1


    # Get the node at a specific index
    def get_node_by_index(self, index):
        if index < 0:
            return None
        p = self.head
        for i in range(index):
            if p is None:
                return None
            p = p.next
        return p

    # Insert node at a specific index
    def add_by_index(self, index, item):
        if index < 0:
            return self.head

        if index == 0:
            return self.add_first(item)

        prev = self.get_node_by_index(index-1)
        if prev:
            self.add_after(prev, item)
        self.size += 1
        return self.head

    # Insert into ordered list
    def add_to_ordered_list(self, item):
        p = self.head
        q = None
        self.size += 1

        while p and p.data < item:
            q = p
            p = p.next

        if q is None:
            return self.add_first(item)
        else:
            return self.add_after(q, item)
            # return self.head
        

    # Find node of value
    def find_node(self, item):
        cur = self.head
        while cur:
            if cur.data == item:
                return cur
            cur = cur.next
        return None

    # Remove first node
    def remove_first(self):
        if self.head is None:
            return None
        temp = self.head
        self.head = self.head.next
        if self.head:
            self.head.prev = None
        self.size -= 1
        return self.head
    
    #Remove last node
    def remove_last(self):
        if self.tail is None:
            return None
        temp = self.tail
        self.tail = self.tail.prev
        if self.tail:
            self.tail.next = None
        else:
            self.head = None
        self.size -= 1
        return self.tail

    # Remove node after a specific node
    def remove_after(self, prv):
        if prv is None or prv.next is None:
            return
        temp = prv.next
        prv.next = temp.next
        if temp.next:
            temp.next.prev = prv
        else:
            self.tail = prv
        self.size -= 1

    # Remove node by index
    def remove_by_index(self, index):
        if index < 0:
            return self.head

        if index == 0:
            return self.remove_first()

        prev = self.get_node_by_index(index-1)
        if prev and prev.next:
            self.remove_after(prev)
        self.size -= 1
        return self.head

    # Remove node by value
    def remove_by_value(self, item):
        p = self.head
        q = None

        while p and p.data != item:
            q = p
            p = p.next

        if p is None:
            return self.head

        if q is None:
            return self.remove_first()
        else:
            self.remove_after(q)
            return self.head

    
    # Function to merge two sorted lists
    @staticmethod
    def merge(list1, list2):
        dummy = Node(0)
        tail = dummy

        while list1 and list2:
            if list1.data <= list2.data:
                tail.next = list1
                list1.prev = tail
                list1 = list1.next
            else:
                tail.next = list2
                list2.prev = tail
                list2 = list2.next
            tail = tail.next

        tail.next = list1 if list1 else list2

        if tail.next:
            tail.next.prev = tail

        return dummy.next

    # Function to concatenate two lists
    # def concat(list1, list2):
    #     if list1 is None:
    #         return list2
    #     if list2 is None:
    #         return list1
    #     p = list1
    #     while p.next:
    #         p = p.next
    #     p.next = list2
    #     list2.prev = p
    #     return list1

    def concat(self, list2):
        if self.head is None:
            self.head = list2.head
            self.tail = list2.tail
            self.size = list2.size
        elif list2.head is None:
            return
        else:
            self.tail.next = list2.head
            list2.head.prev = self.tail
            self.tail = list2.tail
            self.size += list2.size

    # Function to get the middle node
    def get_mid(self):
        slow = self.head
        fast = self.head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        return slow

    # Function to sort the list (ascending order)
    # def sort(self):
    #     if self.head is None or self.head.next is None:
    #         return self.head

    #     mid = self.head.get_mid()
    #     half = mid.next
    #     mid.next = None
    #     if half:
    #         half.prev = None

    #     return self.merge(self.sort(head), self.sort(half))
    

    def sort(self):
        if self.head is None or self.head.next is None:
            return self.head

        mid = self.get_mid()
        half = mid.next
        mid.next = None
        if half:
            half.prev = None

        left_sorted = LList()
        left_sorted.head = self.head
        left_sorted.size = self.size // 2
        right_sorted = LList()
        right_sorted.head = half
        right_sorted.size = self.size - left_sorted.size

        left_sorted = left_sorted.sort()
        right_sorted = right_sorted.sort()
        self.head = self.merge(left_sorted, right_sorted)

        cur = self.head
        while cur and cur.next:
            cur = cur.next
        self.tail = cur

        return self.head
    # Function to sort the list using insertion sort
    def insertion_sort(self):
        if not self.head or not self.head.next:
            return self.head

        sorted_head = self.head
        cur = self.head.next
        sorted_head.next = None

        while cur:
            next_cur = cur.next
            if cur.data < sorted_head.data:
                cur.next = sorted_head
                sorted_head.prev = cur
                sorted_head = cur
            else:
                search = sorted_head
                while search.next and search.next.data < cur.data:
                    search = search.next
                cur.next = search.next
                if search.next:
                    search.next.prev = cur
                search.next = cur
                cur.prev = search
            cur = next_cur

        self.head = sorted_head
        self.tail = sorted_head
        while self.tail and self.tail.next:
            self.tail = self.tail.next

        return self.head

    # Function to sort the list using selection sort
    def selection_sort(self):
        if not self.head:
            return self.head

        cur = self.head
        while cur:
            min_node = cur
            search = cur.next
            while search:
                if search.data < min_node.data:
                    min_node = search
                search = search.next
            if min_node != cur:
                cur.data, min_node.data = min_node.data, cur.data
            cur = cur.next

        return self.head


    # Function to reverse the list
    def reverse(self):
        prev = None
        cur = self.head
        while cur:
            next_node = cur.next
            cur.next = prev
            cur.prev = next_node
            prev = cur
            cur = next_node
        self.tail = self.head
        self.head = prev
        return self.head
    

    def rotate(self, d):
        """
        d%size: 실제로 회전하는 횟수(한바퀴, 즉 d == size인 경우는 제자리임으로)
        d = 0: 아무런 작업도 하지 않음
        d > 0: 오른쪽 회전
        d < 0: 왼쪽 회전
        """
        if self.head == None or d == 0 or self.size == 0:
            return
        d %= self.size #실제 회전수
        if d == 0: #그냥 한바퀴인 경우
            return
        if d < 0:
            d = self.size + d #왼쪽으로 d만큼이라는 것은 오른쪽으로 size-d만큼이라는것

        ##오른쪽 회전
        #1. 새로운 tail 먼저 지정
        new_tail = self.tail
        for _ in range(d):
            new_tail = new_tail.prev
        
        #새로운 head: 새로운 tail 바로 옆
        new_head = new_tail.next
        new_head.prev = None
        new_tail.next = None #연결 끊기

        #원래 끊어진 부분 잇기
        self.head.prev = self.tail
        self.tail.next = self.head

        #head, tail 갱신
        self.head = new_head
        self.tail = new_tail

        # Convert an array to a linked list
    def array_to_linked_list(self, arr):
        self.clear()  # 기존 리스트를 비웁니다.
        for item in arr:
            self.add_last(item)  # 배열의 각 요소를 리스트의 끝에 추가합니다.

    # 추가된 함수: 리스트의 끝에 노드를 추가하는 함수
    def add_last(self, item):
        if not self.head:  # 리스트가 비어 있는 경우
            self.add_first(item)
        else:
            temp = Node(item)
            self.tail.next = temp
            temp.prev = self.tail
            self.tail = temp
            self.size += 1
    
        # Convert the linked list to an array
    def linked_list_to_array(self):
        arr = []
        current = self.head
        while current:
            arr.append(current.data)  # 현재 노드의 데이터를 배열에 추가
            current = current.next
        return arr




if __name__ == "__main__":
    # Create an empty list and check if it's empty
    llist = LList()
    print("List initially empty:", llist.is_empty())

    # Add elements to the list
    llist.add_first(10)
    llist.add_first(20)
    llist.add_first(30)
    print("List after adding elements at the beginning:")
    llist.print_list()

    # Add elements by index
    llist.add_by_index(1, 15)
    llist.add_by_index(0, 25)
    print("List after adding elements by index:")
    llist.print_list()

    # Add to ordered list
    llist.add_to_ordered_list(18)
    llist.add_to_ordered_list(5)
    print("List after adding to ordered list:")
    llist.print_list()

    # Get first and last elements
    print("First element:", llist.get_first())
    print("Last element:", llist.get_last())

    # Find a node
    node = llist.find_node(15)
    print("Finding node with value 15:", "Found" if node else "Not found")

    # Add before a specific node
    node = llist.find_node(18)
    if node:
        llist.add_before(node, 17)
        print("List after adding 17 before 18:")
        llist.print_list()

    # Add after a specific node
    node = llist.find_node(10)
    if node:
        llist.add_after(node, 12)
        print("List after adding 12 after 10:")
        llist.print_list()

    # Remove elements
    llist.remove_by_value(25)
    print("List after removing element with value 25:")
    llist.print_list()

    llist.remove_by_index(2)
    print("List after removing element at index 2:")
    llist.print_list()

    llist.remove_first()
    print("List after removing the first element:")
    llist.print_list()

    llist.remove_last()
    print("List after removing the last element:")
    llist.print_list()

    # Remove after a specific node
    node = llist.find_node(15)
    if node:
        llist.remove_after(node)
        print("List after removing the node after 15:")
        llist.print_list()

    # Sort the list
    llist.insertion_sort()
    print("List after sorting:")
    llist.print_list()

    # Reverse the list
    llist.reverse()
    print("List after reversing:")
    llist.print_list()

    # Rotate the list
    llist.rotate(2)
    print("List after rotating by 2 positions:")
    llist.print_list()

    # Merge with another list
    list2 = LList()
    list2.add_first(40)
    list2.add_first(35)
    list2.add_first(28)
    print("Second list before concate:")
    list2.print_list()


    llist.concat(list2)
    print("List after concatenating second list:")
    llist.print_list()

     # Reset the list
    llist.clear()
    llist.add_first(4)
    llist.add_first(1)
    llist.add_first(3)
    llist.add_first(2)
    llist.add_first(16)
    llist.add_first(9)
    llist.add_first(10)
    llist.add_first(14)
    llist.add_first(8)
    llist.add_first(7)

    # Test selection sort
    llist.selection_sort()
    print("List after selection sort:")
    llist.print_list()

    # Convert linked list to array
    array_from_list = llist.linked_list_to_array()
    print("Array after converting from linked list:", array_from_list)
    # Convert array back to linked list
    new_llist = LList()  # 새로운 연결 리스트 생성
    new_llist.array_to_linked_list(array_from_list)
    print("Linked list after converting back from array:")
    new_llist.print_list()

    # Clear the list
    llist.clear()
    print("List after clearing:")
    print("Size of the list:", llist.get_size())
    llist.print_list()
